dot/parachain/bitfield-distribution/bitfield_distribution.go

```diff
@@ -0,0 +1,533 @@
+	networkbridgemessages "github.com/ChainSafe/gossamer/dot/parachain/network-bridge/messages"
+	provisionermessages "github.com/ChainSafe/gossamer/dot/parachain/provisioner/messages"
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/dot/parachain/util"
+	validationprotocol "github.com/ChainSafe/gossamer/dot/parachain/validation-protocol"
+	"github.com/ChainSafe/gossamer/internal/log"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "parachain-bitfield-distribution"))
+
+type perRelayParentData struct {
+	// Signing context for a particular relay parent.
+	// the required part of the signing context
+	sessionIndex parachaintypes.SessionIndex
+
+	// Set of validators for a particular relay parent.
+	validatorsSet []parachaintypes.ValidatorID
+
+	// Set of validators for a particular relay parent for which we
+	// received a valid `BitfieldGossipMessage`.
+	// Also serves as the list of known messages for peers connecting
+	// after bitfield gossips were already received.
+	onePerValidator map[parachaintypes.ValidatorID]*validationprotocol.BitfieldDistributionMessage
+
+	// Avoid duplicate message transmission to our peers.
+	messageSentToPeer map[peer.ID]map[parachaintypes.ValidatorID]struct{}
+
+	// Track messages that were already received by a peer to prevent flooding.
+	messageReceivedFromPeer map[peer.ID]map[parachaintypes.ValidatorID]struct{}
+}
+
+// this will be used in the PeerViewChange handler so skip the lint check for now
+//
+//nolint:all
+func newPerRelayParentData(sessionIndex parachaintypes.SessionIndex, validatorSet []parachaintypes.ValidatorID,
+) *perRelayParentData {
+	return &perRelayParentData{
+		sessionIndex:            sessionIndex,
+		validatorsSet:           validatorSet,
+		onePerValidator:         make(map[parachaintypes.ValidatorID]*validationprotocol.BitfieldDistributionMessage),
+		messageSentToPeer:       make(map[peer.ID]map[parachaintypes.ValidatorID]struct{}),
+		messageReceivedFromPeer: make(map[peer.ID]map[parachaintypes.ValidatorID]struct{}),
+	}
+}
+
+// messageFromValidatorNeededByPeer determines if that particular message signed by a
+// validator is needed by the given peer.
+func (p *perRelayParentData) messageFromValidatorNeededByPeer(peerID peer.ID, signedBy parachaintypes.
+	ValidatorID) bool {
+	_, sendToExist := p.messageSentToPeer[peerID][signedBy]
+	_, receiveFromExist := p.messageReceivedFromPeer[peerID][signedBy]
+
+	return !sendToExist && !receiveFromExist
+}
+
+type BitfieldDistribution struct {
+	subSystemToOverseer chan<- any
+
+	peerViews map[peer.ID]struct {
+		view            parachaintypes.View
+		protocolVersion uint32
+	}
+	ourView        parachaintypes.View
+	topologies     grid.SessionGridTopologyStorage
+	perRelayParent map[common.Hash]*perRelayParentData
```

Comment:
is this meant to be optional?

Reply:
what you mean by optional? they are init with empty value and will be modified by the corresponding handler

Reply:
I think @timwu20 meant that `*perRelayParentData` being a pointer suggests that the map value could be `nil`.

Reply:
I think pointer here makes more sense because 1) it is used as value of a map which means it exists only with an associated valid RelayParent hash, default(empty) perRelayParentData doesn't really make sense here IMO; 2) it contains bunch of slices and maps, so we probably don't want to copy the entire thing around rather than just provide a reference. Let me know if there is something else that I didn't see.

Reply:
I just saw you do nil checks in other parts of the code, so I thought it was meant to be optional.